### ISSUE TITLE
Extract method can create breaking changes if precompiler directives are used.

### ISSUE THREAD

**Issue by comintern** [rubberduck-vba/Rubberduck#2025]
Not exactly sure what the best way to handle this is but if variables are declared inside a precompiler block, only one set of parameters is generated.  Given this code...

``` VB
Option Explicit

#Const LateBound = True

Private Sub Foobar()
    #If LateBound Then
        Dim Dict As Scripting.Dictionary
        Set Dict = New Scripting.Dictionary
    #Else
        Dim Dict As Object
        Set Dict = CreateObject("Scripting.Dictionary")
    #End If

    Dim Key As String
    Dim Value As String

    Key = "Foo"                  'Selection start.
    Value = "Bar"
    Dict.Add Key, Value
    Debug.Print Dict(Key)   'Selection end.
End Sub
```

...using extract method on the noted selection results in this:

``` VB
Option Explicit

#Const LateBound = True

Private Sub Foobar()
    #If LateBound Then
        Dim Dict As Scripting.Dictionary
        Set Dict = New Scripting.Dictionary
    #Else
        Dim Dict As Object
        Set Dict = CreateObject("Scripting.Dictionary")
    #End If

    Dim Key As String
    Dim Value As String

    NewMethod Dict, Key
End Sub

Private Sub NewMethod(ByRef Dict As Scripting.Dictionary, ByRef Key As String)
    Key = "Foo"
    Value = "Bar"
    Dict.Add Key, Value
    Debug.Print Dict(Key)
End Sub
```

While this is a bit contrived as an example, if the directive is something like `Vba7` and `LongPtr` or `Long` parameters, this would be a breaking change.

**Comment by retailcoder** [comment]
@comintern I think it was broken in the first place:

    #If LateBound Then
        Dim Dict As Scripting.Dictionary
        Set Dict = New Scripting.Dictionary
    #Else
        Dim Dict As Object
        Set Dict = CreateObject("Scripting.Dictionary")
    #End If

Should be

    #If Not LateBound Then
        Dim Dict As Scripting.Dictionary
        Set Dict = New Scripting.Dictionary
    #Else
        Dim Dict As Object
        Set Dict = CreateObject("Scripting.Dictionary")
    #End If

---

Jokes aside, so the problem is that the `Value` variable doesn't get carried into a parameter? I'm not sure I understand what role the conditional compilation directives are playing here.

**Comment by comintern** [comment]
OK, bad example.  In the one above assume that the LateBound is true when there isn't a reference to the MS Scripting Runtime.  More broadly, the issue is that if the type of variable that is being extracted can change, it can break the implementation of the extracted method.

This might be a better example:

Option Explicit

```VB
Public Sub Foo()
#If VBA7 Then
    Dim Bar As LongPtr
    Dim Baz As LongPtr
#Else
    Dim Bar As Long
    Dim Baz As Long
#End If
    Dim ApiVal As String
    Bar = SomeAPIFunction(ApiVal)
    Baz = SomeOtherAPIFunction(Bar)
    Debug.Print Baz
End Sub
```

Extracts to:
```VB
Private Sub NewMethod(ByRef ApiVal As String)
    Dim Bar As LongPtr   '<-- this is a problem.
    Dim Baz As LongPtr   '<-- this is a problem.
    Dim ApiVal As String
    Bar = SomeAPIFunction(ApiVal)
    Baz = SomeOtherAPIFunction(Bar)
    Debug.Print Baz
End Sub
```

**Comment by retailcoder** [comment]
Ah, I see now.

Unfortunately there is no way around this, other than extracting *everything* as `Variant`. The "dead code" simply does not exist as far as Rubberduck knows.

**Comment by grleachman** [comment]
The bigger problem is that the extract does not make use of proper expression trees. I was not able to find a way of extracting an expression tree for a method. If someone could tell me how to, or show me where it is done, then I will gladly rewrite the Refactor method.

That said pre-compiler declarations were never considered when I refactored the code.

So there are 2 views one could take
1) If one used an expression tree builder, then one would rely on the expression tree builder to evaluate definition selection, and therefore the expression tree builder would need to recognise pre-compiler directives.
2) If one could simply consider pre-compiler options. This might not be such a problem, because the scoping of the variable is either local to the method, or global. The variable would then live within a pre-compiler directive, and it would be easy to extract the pre-compilation statement.

Outside of those 2, there is the problem of how would you refactor a method that had a pre-compiler on a variable that then became a parameter. One might have to refactor into 2 separate methods, one for each set of pre-compiler definitions? This might lead to entire blocks of code that were one flavour or another. 

This might open up possibilities for a whole other refactoring opportunity... "Group Pre-Compilations"

    Public Sub Foo()
    #If VBA7 Then
        Dim Bar As LongPtr
        Dim Baz As LongPtr
    #Else
        Dim Bar As Long
        Dim Baz As Long
    #End If
        Bar = SomeAPIFunction(Baz )
        Debug.Print Baz
    End Sub

extracting to 

    Public Sub Foo()
    #If VBA7 Then
        Dim Bar As LongPtr
        Dim Baz As LongPtr
    #Else
        Dim Bar As Long
        Dim Baz As Long
    #End If
        NewMethod(Baz)
    #end
    End Sub

    #If VBA7 Then
         Private Sub NewMethod(ByRef Baz As LongPtr)
            Dim Bar As LongPtr          
            Bar = SomeAPIFunction(Baz )
            Debug.Print Bar
        End Sub  
    #Else
        Private Sub NewMethod(ByRef Baz As Long)
            Dim Bar As Long
            Bar = SomeAPIFunction(Baz)
            Debug.Print Bar
        End Sub  
    #End If

**Comment by comintern** [comment]
@grleachman That would be awesome, but I was thinking of starting with something simpler like a warning to the effect of "Hey, there's a high probability that you're going to break this. Are you sure you want to do that?"

**Comment by grleachman** [comment]
Question:
 - Is the parser ever going to provide a parse tree? I think I remember someone commenting that they had already done this.
 - are compiler directives. #const #if  supported in the parser.. I cannot see any in the ANTLR grammar definitions.

**Comment by Hosch250** [comment]
1. Yes, the parser currently does provide a parse tree.  You can get the parse tree for a module by going into a declaration from that module and accessing the `Context` property.  I can warn you that it is awful to work with.  There is a Roslyn-style tree planned for the future (read easily usable/modifiable), but that is a long ways off still.

2. Check this file: https://github.com/rubberduck-vba/Rubberduck/blob/next/Rubberduck.Parsing/Preprocessing/VBAConditionalCompilationParser.g4

3. It already does parse modules in isolation.  You used to be able to request a parse for any given module, but I removed that option because it was extremely bug-prone when the previous parse was cancelled for a single module to be parsed.  Methods, not yet, and I'm not sure anyone knows how to, but it is planned.

**Comment by retailcoder** [comment]
The problem with parsing an individual procedure is that the token positions in the parse tree no longer match the token positions in the VBE. You could get a parse tree, but not the symbol table derived from it (declarations and identifier references), so I'm not sure how useful a parse tree with offset token positions and no symbol table would be.

**Comment by retailcoder** [comment]
Also, precompiler directives are parsed separately so as to only get a parse tree of the 'live' code. This is an improvement over the original grammar where precompiler directives that interfere with other parser rules simply could not be parsed, and a parse tree that would contain duplicate or conflicting declarations could not be processed.

The only *expression trees* we have are used in the resolver, to determine the nature of `Foo.Bar` (or to evaluate constant expressions) - conditional compilation is not part of this.

As @Hosch250 said, the parser tree context for any given procedure can be accessed from that procedure's declaration's `Context` property, but that is a `ParserRuleContext` object. The actual `IParseTree` for a given module can be accessed from the parser state; as long as Rubberduck uses token positions from the parse trees to position declarations and identifier references, nothing can be accomplished with an individual procedure's parse tree - and even if we could, the precompiler directives would still throw everything off, because none of all the contributors that touched the ANTLR grammar (including the original author) could not get a valid parse tree out of code that includes precompiler directives that interfere with other parser rules.

In other words, there's no possible fix for this; precompiler directives cannot be included in the parse tree processed by the resolver, because they contradict scoping rules and make a given declaration exist in two places or more. I don't think the VBA interpreter sees the dead code conditionally compiled - expecting Rubberduck to do this is not fair IMO, it's a problem we're solving the way VBA solved it - by ignoring the code that wouldn't be compiled. We could tweak things so as to "remember" what code we're ignoring, but the ignored code could not live in a parse tree anyway, because of EBNR limitations.

**Comment by grleachman** [comment]
So what I'm trying to find is a tree. The top node being the method header, all nodes off of the top node are either blocks or expressions. a block is a top node containing blocks or expressions . recursively.


block : if, while, do, etc
expression : call , assignment, etc

validating a method extraction selection would then require confirming that the selection didn't end and star outside of the same block.


given that there is only #if  and #const as pre-compilers...

The first creates a block, and the second creates an expression. So the first is a branching condition, Everything in that block then becomes a member of n conditional branches.

Speculating, but given that we are talking only about an extract method which will be limited in size. If you had the expression tree for the part of the code being extracted, had the original parse object for the pre-extract parse. You could construct the required new code, and the reparse. You don't need exact position for extract method, only relative positions.

I should try it out and get back to you. I need to review what the Context object was. Perhaps I'll try and create a tree from it if possible.

**Comment by Hosch250** [comment]
The `Context` object is a node in the parse tree.  The ANTLR tree is a horrible, horrible pain to work with, which is the main reason I'm looking to make a simpler tree with all the relevant information together.

**Comment by retailcoder** [comment]
@grleachman except you won't find conditional compilation statements in the parse tree or in any procedure's `Context`, because conditional compilation statements are processed separately, *before* we give the parser any code to parse. I agree that only relative positioning would be needed, but it seems overkill to reparse a procedure when we already know everything there is to know about it through its `Declaration` and that declaration's `Context`, which you can walk manually if you like... but as @Hosch250 says it's a pain in the neck to work with - you'd probably rather want to implement a parse tree listener and walk the module's parse tree to do this instead... but then again, until we keep track of what conditionally-compiled code we're filtering out, even that isn't going to tell you when you're entering or exiting a conditional compilation block.

`#If` blocks can appear anywhere and be intertwined between `Function` signatures and anything else: that's not handled by the EBNR grammar. The solution we came up with was to pre-process the code so as to determine which code is "live", and replace the "dead" code with vertical whitespace before we feed it to our parser. Any better ideas welcome.

**Comment by Vogel612** [comment]
If the VBE enforces syntactical correctness within conditional compilation and especially across non-live code we could try to stitch the mess together. I daresay it's significantly too much work though.

The problem with this is that it's basically impossible to put that into a grammar that can handle all valid combinations. It's hard enough to allow line-continuations where they're allowed by the VBE, Conditional Compilation is **significantly** worse. I daresay that [status-by-design] is the only reasonable response to give here

**Comment by Vogel612** [comment]
Since the Extract Method refactoring is in dire need of a rewrite from scratch, all [bug] issues pertaining to it, are in the process of becoming irrelevant. Closing.

**Comment by bclothier** [comment]
@Vogel612  add to epic

**Comment by bclothier** [comment]
ATM, the solution is to disable the functionality altogether if the selection contains any directives. 

If the module as a whole contains any directives that's not within the selection, the dialog will display a warning instead. 

This decision is necessary because at the moment, we are not actually parsing the directives as part of the code. We do parse them early on to help us decide on a "live code" route but nothing more beyond that. Until the day we do update the parser to handle the directives inline with the code, that is the best and safest solution.
